File: sorting/java/Quick_sort.java
import java.util.*;

class Quick_sort {
    public Vector<Integer> add_vec(Vector<Integer> v1, Vector<Integer> v2){ //function for adding two vectors
        Collection c = v2.subList(0, v2.size());
        v1.addAll(c);
        return v1;
    }

    public Vector<Integer> splice(Vector<Integer> v1,int from,int till){//function for splicing a vector
        Collection c = v1.subList(from, till);
        Vector<Integer> spliced = new Vector<Integer>();
        spliced.addAll(c);
        return spliced;
    }

    public Vector<Integer> quick_sort(Vector<Integer> list){//actual quick sort function
        int n = list.size();
        if(n<2){return list;}           //base case 0 or 1 element, can be another sort algo
        int mid = (int)(Math.random()*n);//picking a random pivot
        int head = 0;
        int tail = n-1;
        int p = list.get(mid);
        while(head<tail){//partition
            while ((list.get(head)<p || head==mid) && head<tail){//searching for an element in LHS >= pivot
                head++;
            }
            while((list.get(tail)>p || tail==mid) && head<tail){//searching for an element in LHS <= pivot
                tail--;
            }
            if(head == tail){ break; }
            int temp = list.get(tail);      //swap
            list.set(tail, list.get(head));
            list.set(head, temp);
            if(tail - head >1){
                tail--;
                head++;
            }else if(head<mid){
                head++;
            }else if(head>mid){
                tail--;
            }
        }
        if(mid<head && list.get(head)>=p){head--;}
        else if(mid>head && list.get(head)<=p){head++;}
        int temp = list.get(mid);                       //placing the pivot at it's sorted position
        list.set(mid, list.get(head));
        list.set(head, temp);
        Vector<Integer> v1 = quick_sort(splice(list, 0, head));
        v1.add(p);
        if(head == n-1){
            return v1;
        }
        Vector v2 = quick_sort(splice(list, head+1, n));
        add_vec(v1, v2);
        return v1;
    }
}

File: sorting/cpp/quick_sort.cpp
#include <vector>
#include <cstdlib>      //required for random number
using namespace std;

vector<int> quick_sort(vector<int> list){
    int n = list.size();
    if(n<2){return list;}       //base case 0 or 1 element, can be another sort algo
    int mid = rand()%n;         //picking a random pivot
    int head = 0;
    int tail = n-1;
    int p = list[mid];
    while (head<tail){  //partition
        while ((list[head]<p || head == mid) && head<tail){//searching for an element in LHS >= pivot
            head++;
        }
        while((list[tail]>p || tail==mid) && head<tail){//searching for an element in LHS <= pivot
            tail--;
        }
        if(head == tail){ break; }
        int temp = list[tail];          //swap
        list[tail] = list[head];
        list[head] = temp;
        if(tail - head >1){
            tail--;
            head++;
        }else if(head<mid){
            head++;
        }else if(head>mid){
            tail--;
        }
    }
    if(mid<head && list[head]>=p){ head--;}
    else if(mid>head && list[head]<=p){ head++;}
    int temp=list[mid];                 //placing the pivot at it's sorted position
    list[mid] = list[head];
    list[head]=temp;
    if(head == n-1){
        vector<int> v1 = quick_sort(vector<int>(list.begin(), list.begin()+head));
        v1.push_back(p);
        return v1;
    }
    vector<int> v1 = quick_sort(vector<int>(list.begin(), list.begin()+head));
    vector<int> v2 = quick_sort(vector<int>(list.begin()+head+1, list.end()));
    v1.push_back(p);
    v1.insert(v1.end(), v2.begin(), v2.end() );
    return v1;
}